README.md
# SPIDAM
 
This program measures and analyzes reverberation time in a given audio file (if the file is not .wav, it will be converted to .wav) and plots the data as multiple different graph, as well as displaying the spectogram of the file.

This program exists so that users can graph audio files and use the data as needed.

## Usage

When you run the controller.py program a new GUI will open. Make sure you downloaded the requirements from requirements.txt.

Click on "Load Audio File" and your files will be opened. Click on the audio file you want to graph and click "Open" on the bottom right of the window. When the file loads, a waveform graph will appear. Clicking on the other three buttons will cause other graphs to appear. The "Plot RT60" button will cause low, mid, and high frequency graphs to appear. The "Merge Graphs" button will cause the low, mid, and high frequencies graphs to merge into one graph. Finally, the "Show Spectogram" button will cause a spectrogram of your file to appear. During the testing of our program, we had some errors displaying the plots the first time a file was loaded. Reloading the file or loading a different file will fix the issue. Also, the spectogram does not work for all audio files.


CHANGELOG.md
0.0.1 Build: Created repository
1.0.0 Build: Program plots waveform graph
1.0.1 Build: Program plots waveform graph, displays name of file, and greatest resonance
1.1.0 Build: Program plots RT60 graph for Low, Mid, High frequencies
1.1.1 Build: Program can merge all three plots
1.1.2 Build: Program displays RT60 difference
1.2.0 Build: Spectrogram button/function added. Program meets all requirements.
1.2.1 Build: Uses a function so that SciPy is no longer needed in IDE to view graphs. Function is in line 8 of view.py and can be removed to return to 1.2.0 build.


program.py
import tkinter as tk
from tkinter import filedialog
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.io import wavfile
import librosa
import librosa.display

class AudioAnalyzer:
    def __init__(self, root):
        self.root = root
        self.root.title("Audio Analyzer")

        self.load_button = tk.Button(root, text="Load Audio File", command=self.load_audio)
        self.load_button.pack(pady=10)

        self.result_label = tk.Label(root, text="")
        self.result_label.pack()

    def load_audio(self):
        file_path = filedialog.askopenfilename(title="Select Audio File", filetypes=[("Audio files", "*.wav;*.mp3;*.aac")])

        if file_path:
            self.process_audio(file_path)

    def process_audio(self, file_path):
        self.result_label.config(text="Processing...")

        # Check if the file is WAV, if not, convert it
        if not file_path.lower().endswith('.wav'):
            converted_path = self.convert_to_wav(file_path)
            if converted_path:
                file_path = converted_path
            else:
                return

        # Display file name
        self.result_label.config(text=f"File: {os.path.basename(file_path)}")

        # Load WAV file
        sample_rate, audio_data = wavfile.read(file_path)

        # Check for meta data (tags) and remove if present
        if 'INFO' in audio_data:
            audio_data = audio_data['INFO']

        # Check for multi-channel and convert to one channel
        if len(audio_data.shape) > 1:
            audio_data = audio_data[:, 0]

        # Display time value of .wav in seconds
        duration = len(audio_data) / sample_rate
        print(f"Duration of audio: {duration} seconds")

        # Display waveform of .wav
        plt.figure(figsize=(10, 4))
        librosa.display.waveshow(audio_data, sr=sample_rate)
        plt.title('Waveform')
        plt.show()

        # Data Analysis and Modeling (example using NumPy and SciPy)
        rt60_low, rt60_mid, rt60_high = self.calculate_rt60(audio_data, sample_rate)

        print(f"RT60 Low: {rt60_low} seconds")
        print(f"RT60 Mid: {rt60_mid} seconds")
        print(f"RT60 High: {rt60_high} seconds")

        # Visualize RT60 data for each frequency range
        self.plot_rt60(rt60_low, rt60_mid, rt60_high)

    def convert_to_wav(self, file_path):
        try:
            y, sr = librosa.load(file_path, sr=None)
            new_path = file_path.rsplit('.', 1)[0] + '.wav'
            wavfile.write(new_path, sr, y)
            return new_path
        except Exception as e:
            print(f"Error converting to WAV: {e}")
            return None

    def calculate_rt60(self, audio_data, sample_rate):
        # Calculate RT60 for Low, Mid, and High frequencies using LibROSA
        # You may need to adjust this based on your specific requirements

        # Compute Short-Time Fourier Transform (STFT)
        stft = librosa.stft(audio_data)

        # Calculate power spectrogram
        power_spec = np.abs(stft) ** 2

        # Define frequency bands (Low, Mid, High)
        freq_bins = librosa.fft_frequencies(sr=sample_rate)
        low_freqs = (freq_bins >= 20) & (freq_bins < 500)
        mid_freqs = (freq_bins >= 500) & (freq_bins < 2000)
        high_freqs = (freq_bins >= 2000) & (freq_bins <= 8000)

        # Integrate power within each frequency band
        power_low = np.sum(power_spec[low_freqs, :])
        power_mid = np.sum(power_spec[mid_freqs, :])
        power_high = np.sum(power_spec[high_freqs, :])

        # Calculate RT60 for each frequency band
        rt60_low = self.compute_rt60(power_low, sample_rate)
        rt60_mid = self.compute_rt60(power_mid, sample_rate)
        rt60_high = self.compute_rt60(power_high, sample_rate)

        return rt60_low, rt60_mid, rt60_high

    def compute_rt60(self, power, sample_rate):
        # Compute RT60 from power spectrogram
        # You may need to adjust this based on your specific requirements

        # Sum the power over time
        power_sum = np.sum(power, axis=1)

        # Find the time point where the power drops to half
        half_power = 0.5 * power_sum[-1]
        t_half = np.argmax(power_sum > half_power)

        # Calculate RT60 in seconds
        rt60 = 2 * (len(power_sum) - t_half) / sample_rate

        return rt60

    def plot_rt60(self, rt60_low, rt60_mid, rt60_high):
        # Plot RT60 for Low, Mid, and High frequencies
        frequencies = ['Low', 'Mid', 'High']
        rt60_values = [rt60_low, rt60_mid, rt60_high]

        plt.bar(frequencies, rt60_values, color=['blue', 'green', 'red'])
        plt.title('RT60 for Low, Mid, High Frequencies')
        plt.xlabel('Frequency Range')
        plt.ylabel('RT60 (seconds)')
        plt.show()


if __name__ == "__main__":
    root = tk.Tk()
    app = AudioAnalyzer(root)
    root.mainloop()


controller.py
# controller.py
from model import AudioModel
from view import AudioView
import tkinter as tk
from tkinter import filedialog
from scipy.io import wavfile
import librosa
import numpy as np
import os
import subprocess
import matplotlib.pyplot as plt
from pydub import AudioSegment


class AudioController:
    def __init__(self, root):
        self.model = AudioModel()
        self.view = AudioView(root, self)

    def load_audio(self):
        file_path = filedialog.askopenfilename(title="Select Audio File",
                                               filetypes=[("Audio files", "*.wav;*.mp3;*.aac")])

        if file_path:
            self.model.set_file_path(file_path)
            self.view.display_file_name(os.path.basename(file_path))

            sample_rate, audio_data = self.load_audio_file(file_path)
            self.model.set_audio_data(sample_rate, audio_data)
            self.view.display_waveform(audio_data, sample_rate)

            rt60_low, rt60_mid, rt60_high = self.calculate_rt60(audio_data, sample_rate)
            self.model.set_rt60_values(rt60_low, rt60_mid, rt60_high)

    def load_audio_file(self, file_path):
        if not file_path.lower().endswith('.wav'):
            ffmpeg_path = os.path.join(os.path.dirname(__file__), 'ffmpeg.exe')
            output_path = os.path.join(os.path.dirname(file_path), 'output.wav')

            # Use ffmpeg to convert the input file to WAV and remove metadata
            command = [ffmpeg_path, '-i', file_path, '-acodec', 'pcm_s16le', '-ar', '44100', '-ac', '1', output_path]
            subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)

            # Load the processed WAV file
            audio = AudioSegment.from_file(output_path)
            channels = audio.channels
            if channels > 1:
                # Convert to mono (single channel) if there are multiple channels
                audio = audio.set_channels(1)
            audio_data = np.array(audio.get_array_of_samples())

            # Remove the temporary output file
            os.remove(output_path)

            return audio.frame_rate, audio_data
        else:
            return wavfile.read(file_path)


    def load_wav(self, file_path):
        if not file_path.lower().endswith('.wav'):
            y, sr = librosa.load(file_path, sr=None)  # Load audio
            y = librosa.to_mono(y)  # Convert to mono if it's stereo
            y = librosa.util.normalize(y.astype(float))  # Ensure floating-point format
            return sr, y
        else:
            return wavfile.read(file_path)

    def calculate_rt60(self, audio_data, sample_rate):
        # Compute autocorrelation
        autocorr = np.correlate(audio_data, audio_data, mode='full')

        # Get frequency bands
        low_freqs, mid_freqs, high_freqs = self.get_frequency_ranges(sample_rate)

        # Compute RT60 for each frequency range
        rt60_low = self.compute_rt60(autocorr[low_freqs], sample_rate)
        rt60_mid = self.compute_rt60(autocorr[mid_freqs], sample_rate)
        rt60_high = self.compute_rt60(autocorr[high_freqs], sample_rate)


        return rt60_low, rt60_mid, rt60_high

    def compute_rt60(self, autocorr, sample_rate):
        # Compute power and sum along the appropriate axis
        power = np.abs(np.fft.fft(autocorr, axis=0))

        if power.ndim == 1:
            power_sum = np.sum(power)
            t_half = np.argmax(power > 0.5 * power_sum)
            rt60 = 2 * (len(autocorr) - t_half) / sample_rate
        else:
            power_sum = np.sum(power, axis=1)
            t_half = np.argmax(power_sum > 0.5 * power_sum[-1])
            rt60 = 2 * (len(autocorr) - t_half) / sample_rate

        rt60_array = 2 * (len(autocorr) - np.arange(len(autocorr))) / sample_rate

        return rt60_array

    def get_frequency_ranges(self, sample_rate):
        low_freqs = np.arange(0, 1000)  # Adjust the upper limit as needed
        mid_freqs = np.arange(1000, 5000)  # Adjust the upper limit as needed
        high_freqs = np.arange(5000, sample_rate / 2)  # Adjust the lower and upper limits as needed

        # Convert high_freqs to integers
        high_freqs = high_freqs.astype(int)

        return low_freqs, mid_freqs, high_freqs

    def plot_rt60(self):
        rt60_low = np.atleast_1d(self.model.rt60_low)
        rt60_mid = np.atleast_1d(self.model.rt60_mid)
        rt60_high = np.atleast_1d(self.model.rt60_high)



        min_len = min(rt60_low.size, rt60_mid.size, rt60_high.size)

        rt60_low = rt60_low[:min_len]
        rt60_mid = rt60_mid[:min_len]
        rt60_high = rt60_high[:min_len]

        self.model.set_rt60_values(rt60_low, rt60_mid, rt60_high)

        self.view.display_rt60_values(np.array(rt60_low), np.array(rt60_mid), np.array(rt60_high),
                                      self.merge_rt60_values, self.unmerge_rt60_values)

    def merge_rt60_values(self, rt60_low, rt60_mid, rt60_high):
        # Create a new figure for the merged graph
        plt.figure(figsize=(10, 4))

        # Plot the merged graph
        plt.plot(rt60_low, label='Low Frequency', color='blue')
        plt.plot(rt60_mid, label='Mid Frequency', color='green')
        plt.plot(rt60_high, label='High Frequency', color='red')

        plt.title('Merged RT60 for Low, Mid, High Frequencies')
        plt.xlabel('Time (seconds)')
        plt.ylabel('Power (dB)')
        plt.legend()

        # Calculate the RT60 difference
        avg_rt60 = np.mean([np.mean(rt60_low), np.mean(rt60_mid), np.mean(rt60_high)])
        rt60_difference = avg_rt60 - 0.5

        plt.text(0.5, 0.1, f'RT60 Difference: {rt60_difference:.2f} seconds', transform=plt.gca().transAxes,
                 color='black', fontsize=12, fontweight='bold', ha='center')

        plt.show()

    def compute_highest_resonance(self, audio_data, sample_rate):
        # Compute the FFT
        fft_result = np.fft.fft(audio_data)

        frequencies = np.fft.fftfreq(len(fft_result), d=1 / sample_rate)

        max_amplitude_index = np.argmax(np.abs(fft_result[1:])) + 1

        return frequencies[max_amplitude_index]

    def unmerge_rt60_values(self, rt60_low, rt60_mid, rt60_high):
        self.view.display_rt60_values(rt60_low, rt60_mid, rt60_high, self.merge_rt60_values, self.unmerge_rt60_values)

    def get_audio_duration(self):
        # Check if audio_data is available in the model
        if self.model.audio_data is not None:
            audio_segment = AudioSegment(
                data=self.model.audio_data.tobytes(),
                sample_width=self.model.audio_data.dtype.itemsize,
                frame_rate=self.model.sample_rate,
                channels=1  # Assuming mono audio
            )
            return audio_segment.duration_seconds
        else:
            return 0

    def calculate_and_display_rt60_difference(self):
        # Get the RT60 values from the model
        rt60_low = np.atleast_1d(self.model.rt60_low)
        rt60_mid = np.atleast_1d(self.model.rt60_mid)
        rt60_high = np.atleast_1d(self.model.rt60_high)

        # Take the average of the RT60 values
        avg_rt60 = np.mean([np.mean(rt60_low), np.mean(rt60_mid), np.mean(rt60_high)])

        # Calculate the difference from the optimum reverb time (0.5 seconds)
        rt60_difference = avg_rt60 - 0.5

        self.view.display_rt60_values(rt60_low, rt60_mid, rt60_high, self.merge_rt60_values, self.unmerge_rt60_values,
                                      rt60_difference)

    def show_spectrogram(self):
        if self.model.audio_data is not None and self.model.sample_rate is not None:
            audio_data = self.model.audio_data
            sample_rate = self.model.sample_rate
            self.view.show_spectrogram(audio_data, sample_rate)
        else:
            print("Error: No audio data available.")


if __name__ == "__main__":
    root = tk.Tk()
    root.title("Scientific Python Interactive Data Acoustic Modeling")

    root.geometry("400x200")

    app = AudioController(root)
    root.mainloop()

view.py
# view.py
import tkinter as tk
import numpy as np
import matplotlib.pyplot as plt
import librosa.util
from matplotlib.backends.backend_tkagg import FigureCanvasTkAgg
from scipy.signal import spectrogram
plt.switch_backend("TkAgg")

class AudioView:
    def __init__(self, root, controller):
        self.root = root
        self.controller = controller

        self.load_button = tk.Button(root, text="Load Audio File", command=self.controller.load_audio)
        self.load_button.pack(pady=10)

        self.result_label = tk.Label(root, text="")
        self.result_label.pack()

        self.plot_button = tk.Button(root, text="Plot RT60", command=self.controller.plot_rt60)
        self.plot_button.pack(pady=10)

        #Merge Button
        self.merge_button = tk.Button(root, text="Merge Graphs", command=lambda: None)
        self.merge_button.pack(pady=10)

        #Spectrogram Button
        self.spectrogram_button = tk.Button(root, text="Show Spectrogram", command=self.controller.show_spectrogram)
        self.spectrogram_button.pack(pady=10)

        self.spectrogram_frame = tk.Frame(root)
        self.spectrogram_frame.pack()

        # Initialize the label for RT60 difference
        self.initialize_rt60_difference_label()


    def display_file_name(self, file_name):
        self.result_label.config(text=f"File: {file_name}")

    def display_waveform(self, audio_data, sample_rate):
        # Convert audio_data to floating-point format
        audio_data_float = librosa.util.normalize(audio_data.astype(float))

        # Plot the waveform with units on axes
        time = np.arange(0, len(audio_data_float)) / sample_rate  # Time in seconds
        plt.figure(figsize=(10, 4))
        plt.plot(time, audio_data_float)
        plt.title('Waveform')
        plt.xlabel('Time (seconds)')
        plt.ylabel('Amplitude')

        # Compute the highest resonance frequency
        highest_resonance_freq = self.controller.compute_highest_resonance(audio_data, sample_rate)

        # Add text for highest resonance frequency
        plt.text(0.5, 0.92, f'Highest Resonance: {highest_resonance_freq:.2f} Hz', transform=plt.gca().transAxes,
                 color='black', fontsize=12, fontweight='bold', ha='center')

        # Add text for time duration
        audio_duration = len(audio_data_float) / sample_rate
        plt.text(0.5, 0.85, f'Duration: {audio_duration:.2f} seconds', transform=plt.gca().transAxes,
                 color='black', fontsize=12, fontweight='bold', ha='center')

        plt.subplots_adjust(bottom=0.2)

        plt.show()

    def compute_highest_resonance(self, audio_data, sample_rate):
        # Compute the Fast Fourier Transform (FFT)
        fft_result = np.fft.fft(audio_data)

        frequencies = np.fft.fftfreq(len(fft_result), d=1 / sample_rate)

        max_amplitude_index = np.argmax(np.abs(fft_result[1:])) + 1

        return frequencies[max_amplitude_index]

    def display_rt60_values(self, rt60_low, rt60_mid, rt60_high, merge_function, unmerge_function,rt60_difference=None):
        # Convert scalar values to arrays
        rt60_low = np.atleast_1d(rt60_low)
        rt60_mid = np.atleast_1d(rt60_mid)
        rt60_high = np.atleast_1d(rt60_high)

        # Determine the minimum length
        min_len = min(len(rt60_low), len(rt60_mid), len(rt60_high))

        # Get the duration of the audio file
        audio_duration = self.controller.get_audio_duration()

        # Create subplots for the three individual graphs
        fig, axs = plt.subplots(3, 1, figsize=(10, 12))

        time = np.linspace(0, audio_duration, min_len, endpoint=False)  # Use endpoint=False

        # Plot individual graphs
        axs[0].plot(time, rt60_low[:min_len], label='Low Frequency', color='blue')
        axs[1].plot(time, rt60_mid[:min_len], label='Mid Frequency', color='green')
        axs[2].plot(time, rt60_high[:min_len], label='High Frequency', color='red')

        # Set y-axis limits to better visualize short RT60 values
        axs[0].set_ylim(0, 0.1)
        axs[1].set_ylim(0, 0.1)
        axs[2].set_ylim(0, 0.1)

        axs[0].set_title('RT60 for Low Frequency')
        axs[0].set_xlabel('Time (seconds)')
        axs[0].set_ylabel('Power (dB)')
        axs[0].legend()

        axs[1].set_title('RT60 for Mid Frequency')
        axs[1].set_xlabel('Time (seconds)')
        axs[1].set_ylabel('Power (dB)')
        axs[1].legend()

        axs[2].set_title('RT60 for High Frequency')
        axs[2].set_xlabel('Time (seconds)')
        axs[2].set_ylabel('Power (dB)')
        axs[2].legend()

        plt.tight_layout()

        self.merge_button.config(command=lambda: merge_function(rt60_low, rt60_mid, rt60_high))

        # Display RT60 difference if available
        if rt60_difference is not None:
            # Add text for RT60 difference
            plt.text(0.5, 0.78, f'RT60 Difference: {rt60_difference:.2f} seconds', transform=plt.gca().transAxes,
                     color='black', fontsize=12, fontweight='bold', ha='center')

        plt.show()


    def merge_rt60_values(self, rt60_low, rt60_mid, rt60_high):
        # Create time array for x-axis
        time = np.arange(0, len(rt60_low))

        # Create a new figure for the merged graph
        plt.figure(figsize=(10, 4))

        # Plot the merged graph
        plt.plot(time, rt60_low, label='Low Frequency', color='blue')
        plt.plot(time, rt60_mid, label='Mid Frequency', color='green')
        plt.plot(time, rt60_high, label='High Frequency', color='red')

        plt.title('Merged RT60 for Low, Mid, High Frequencies')
        plt.xlabel('Time (seconds)')
        plt.ylabel('Power (dB)')
        plt.legend()

        plt.show()

    def display_rt60_difference(self, rt60_difference):
        rt60_difference_text = f"Difference: {rt60_difference:.2f} seconds"
        self.rt60_difference_label.config(text=rt60_difference_text)

    def initialize_rt60_difference_label(self):
        self.rt60_difference_label = tk.Label(self.root, text="")
        self.rt60_difference_label.pack()

    def show_spectrogram(self, audio_data, sample_rate):
        # Compute the spectrogram
        f, t, Sxx = spectrogram(audio_data, fs=sample_rate)

        fig, ax = plt.subplots(figsize=(10, 4))
        ax.pcolormesh(t, f, 10 * np.log10(Sxx), shading='auto')
        ax.set_title('Spectrogram')
        ax.set_xlabel('Time (seconds)')
        ax.set_ylabel('Frequency (Hz)')
        ax.figure.colorbar(ax.collections[0], label='Power/Frequency (dB/Hz)')

        canvas = FigureCanvasTkAgg(fig, master=self.spectrogram_frame)
        canvas.draw()
        canvas.get_tk_widget().pack()

        plt.show()
